Sources/Basic/GraphAlgorithms.swift

```diff
@@ -98,3 +98,47 @@ public func topologicalSort<T: Hashable>(
     
     return result.reversed()
 }
+
+/// Finds the first cycle encountered in a graph.
+///
+/// This method uses DFS to look for a cycle and immediately returns when a
+/// cycle is encounted.
+///
+/// - Parameters:
+///   - nodes: The list of input nodes to sort.
+///   - successors: A closure for fetching the successors of a particular node.
+///
+/// - Returns: An array containing nodes of cycle from begining to end and nil if cycle is not found.
+public func findCycle<T: Hashable>(
+    _ nodes: [T], successors: (T) throws -> [T]
+) rethrows -> [T]? {
+    // Ordered set to hold the current traversed path.
+    var path = OrderedSet<T>()
+
+    // Function to visit nodes recursively.
+    // FIXME: Convert to stack.
+    func visit(_ node: T, _ successors: (T) throws -> [T]) rethrows -> [T]? {
+        // If this node is already in the current path then we have found a cycle.
+        if !path.append(node) {
+            let cycle = path[path.index(of: node)!..<path.endIndex]
+            return Array(cycle)
```

Comment:
It is probably good to return the full path to the cycle, so clients can know how to get to the cycle. It could also return two parts, the path to the start and the cycle itself.